xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,246 @@
+
+    next (tok: token, state: RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = DecInData { symbol: bits[SYMBOL_WIDTH]:0, count: bits[COUNT_WIDTH]:0, last: false };
+        let empty = state.count == bits[COUNT_WIDTH]:0;
+        let (input_tok, input) = recv_if(tok, input_r, empty, zero_input);
+        let (next_symbol, next_count, next_last) = if (empty) {
+            let t_count = input.count - bits[COUNT_WIDTH]:1;
+            (input.symbol, t_count, input.last)
+        } else {
+            let t_count = state.count - bits[COUNT_WIDTH]:1;
+            (state.symbol, t_count, state.last)
+        };
+        let send_last = next_last & (next_count == bits[COUNT_WIDTH]:0);
+        let data_tok = send(input_tok, output_s, DecOutData {symbol: next_symbol, last: send_last});
+        RLEDecState {
+            symbol: next_symbol,
+            count: next_count,
+            last: next_last,
+        }
+    }
+}
+
+
+// RLE decoder specialization for the codegen
+proc RLEDec32 {
+    init {()}
+
+    config (
+        input_r: chan<DecInData<32, 2>> in,
+        output_s: chan<DecOutData<32>> out,
+    ) {
+        spawn RLEDec<u32:32, u32:2>(input_r, output_s);
+        ()
+    }
+
+    next (tok: token, state: ()) {
+        ()
+    }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:2;
+
+type TestSymbol = bits[TEST_SYMBOL_WIDTH];
+type TestCount  = bits[TEST_COUNT_WIDTH];
+type TestDecInData  = DecInData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+type TestDecOutData = DecOutData<TEST_SYMBOL_WIDTH>;
+
+
+//const TEST_SEND_LOOKUP = TestDecInData[10]:[
+//    TestDecInData {symbol: TestSymbol:0x1, count: TestCount:0x1, last:true},
+//    TestDecInData {symbol: TestSymbol:0x2, count: TestCount:0x1, last:true},
+//];
+//
+//const TEST_RECV_LOOKUP = TestDecOutData[20]:[
+//    TestDecOutData {symbol: TestSymbol:0x1, last: true},
+//    TestDecOutData {symbol: TestSymbol:0x2, last: true},
+//];
+
+// main proc used to test the RLE decoder
+#[test_proc]
+proc RLEDecTightTester {
+    terminator: chan<bool> out;            // test termination request
+    dec_input_s: chan<TestDecInData> out;
+    dec_output_r: chan<TestDecOutData> in;
+
+    init {()}
+
+    config(terminator: chan<bool> out) {
+        let (dec_input_s, dec_input_r)   = chan<TestDecInData>;
+        let (dec_output_s, dec_output_r) = chan<TestDecOutData>;
+
+        spawn RLEDec<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>(dec_input_r, dec_output_s);
+        (terminator, dec_input_s, dec_output_r)
+    }
+
+    next(tok: token, state: ()) {
+
+        // Simple transaction without repeats
+        let trans_send: (TestSymbol, TestCount)[2] =
+            [(TestSymbol:0xA, TestCount:0x3), (TestSymbol:0xB, TestCount:0x1)];
+        let tok = for ((counter, (symbol_in, count_in)), tok): ((u32, (TestSymbol, TestCount)) , token) in enumerate(trans_send) {
+            let _last = counter == (array_size(trans_send) - u32:1);
```

Comment:
curious why you have `_` here?

Reply:
It' left over from when I had there name collision.
I can remove it, but it will lead to `last: last`

Reply:
I think `last: last` is ok.

Reply:
I've changed all `_last` to `last`